### PR TITLE
Dropping 0.4

### LINKED ISSUE #361: Drop 0.4 support?
I think it would be fairly unusual for packages to be targeting 0.4, 0.5, 0.6, and 0.7 simultaneously. Is it time to drop 0.4 support?

### PR DESCRIPTION
**Pull request by rofinn**
I'll leave updating the test cases till after this has been reviewed to minimize breaking existing functionality.

Fixes #361

### CODE REVIEW

**Comment by stevengj** [comment]
Don't forget to remove things from the README that are no longer needed for v0.5+.

File: README.md [hunk 1]
```diff
@@ -119,8 +94,6 @@ Currently, the `@compat` macro supports the following syntaxes:
 
 * `foreach`, similar to `map` but when the return value is not needed ([#13744])
```
Comment (stevengj) [inline README.md]: Foreach (https://github.com/JuliaLang/julia/pull/13774) was added in 0.5 too.

File: README.md [hunk 2]
```diff
@@ -119,8 +94,6 @@ Currently, the `@compat` macro supports the following syntaxes:
...
 
-* `walkdir`, returns an iterator that walks the directory tree of a directory ([#13707])
-
 * `allunique`, checks whether all elements in an iterable appear only once ([#15914])
```
Comment (stevengj) [inline README.md]: `allunique` was added in 0.5

File: README.md [hunk 3]
```diff
@@ -119,8 +94,6 @@ Currently, the `@compat` macro supports the following syntaxes:
...
 
 * `Base.promote_eltype_op` is available as `Compat.promote_eltype_op`
```
Comment (stevengj) [inline README.md]: added in 0.5

File: README.md [hunk 4]
```diff
@@ -178,37 +151,12 @@ Currently, the `@compat` macro supports the following syntaxes:
...
-
-* `get_rounding` is now `rounding`. `set_rounding` and `with_rounding` are now `setrounding` ([#13232])
-
-*  `Base.tty_size` (which was not exported) is now `displaysize` in Julia 0.5
-
-* `Compat.LinAlg.checksquare` ([#14601])
-
-* `issym` is now `issymmetric` ([#15192])
-
 * `istext` is now `istextmime` ([#15708])
```
Comment (stevengj) [inline README.md]: added in 0.5

File: README.md [hunk 5]
```diff
@@ -178,37 +151,12 @@ Currently, the `@compat` macro supports the following syntaxes:
...
 
-* `symbol` is now `Symbol` ([#16154])
-
 * `write(::IO, ::Ptr, len)` is now `unsafe_write` ([#14766])
```
Comment (stevengj) [inline README.md]: added in 0.5

File: README.md [hunk 6]
```diff
@@ -178,37 +151,12 @@ Currently, the `@compat` macro supports the following syntaxes:
...
 
-* `slice` is now `view` ([#16972]); do `import Compat.view` and then use `view` normally without the `@compat` macro
-
-* `fieldoffsets` has been deprecated in favor of `fieldoffset` ([#14777])
-
 * `print_escaped` is now another method of `escape_string`, `print_unescaped` a method of `unescape_string`, and `print_joined` a method of `join` ([#16603])
```
Comment (stevengj) [inline README.md]: added in 0.5

File: README.md [hunk 7]
```diff
@@ -178,37 +151,12 @@ Currently, the `@compat` macro supports the following syntaxes:
...
 
 * `writemime` has been merged into `show` ([#16563]). Note that to extend this function requires `@compat`; see the [Supported Syntax](#supported-syntax) section for more information
```
Comment (stevengj) [inline README.md]: renamed in 0.5

File: README.md [hunk 8]
```diff
@@ -221,8 +169,6 @@ Currently, the `@compat` macro supports the following syntaxes:
 
 ## New macros
 
-* `@static` has been added ([#16219])
-
 * `@functorize` (not present in any Julia version) takes a function (or operator) and turns it into a functor object if one is available in the used Julia version. E.g. something like `mapreduce(Base.AbsFun(), Base.MulFun(), x)` can now be written as `mapreduce(@functorize(abs), @functorize(*), x)`, and `f(::Base.AbsFun())` as `f(::typeof(@functorize(abs)))`, to work across different Julia versions. `Func{1}` can be written as `supertype(typeof(@functorize(abs)))` (and so on for `Func{2}`), which will fall back to `Function` on Julia 0.5.
```
Comment (stevengj) [inline README.md]: `@functorize` is obsolete in 0.5

File: README.md [hunk 9]
```diff
@@ -221,8 +169,6 @@ Currently, the `@compat` macro supports the following syntaxes:
...
 
 * `Compat.@blasfunc` makes functionality of `Base.LinAlg.BLAS.@blasfunc` available on older Julia versions
```
Comment (stevengj) [inline README.md]: added in 0.5

File: README.md [hunk 10]
```diff
@@ -241,32 +187,18 @@ Currently, the `@compat` macro supports the following syntaxes:
...
-
-* `Base.FS` is now `Base.Filesystem` ([#12819])
-  Compat provides an unexported `Compat.Filesystem` module that is aliased to
-  `Base.FS` on Julia 0.4 and `Base.Filesystem` on Julia 0.5.
-
-* `cov` and `cor` don't allow keyword arguments anymore. Loading Compat defines compatibility methods for the new API ([#13465])
-
 * On versions of Julia that do not contain a Base.Threads module, Compat defines a Threads module containing a no-op `@threads` macro.
 
 * `Base.SingleAsyncWork` is now `Base.AsyncCondition`
```
Comment (stevengj) [inline README.md]: added in 0.5

File: README.md [hunk 11]
```diff
@@ -241,32 +187,18 @@ Currently, the `@compat` macro supports the following syntaxes:
...
   Compat provides an unexported `Compat.AsyncCondition` type that is aliased to
-  `Base.SingleAsyncWork` on Julia 0.4 and `Base.AsyncCondition` on Julia 0.5.
-
-* `repeat` now accepts any `AbstractArray` ([#14082]): `Compat.repeat` supports this new API on Julia 0.4, and calls `Base.repeat` on 0.5.
+  `Base.AsyncCondition` on Julia 0.5.
 
 * `OS_NAME` is now `Sys.KERNEL`. OS information available as `is_apple`, `is_bsd`, `is_linux`, `is_unix`, and `is_windows` ([#16219])
```
Comment (stevengj) [inline README.md]: added in 0.5

File: README.md [hunk 12]
```diff
@@ -241,32 +187,18 @@ Currently, the `@compat` macro supports the following syntaxes:
...
 
 * `cholfact`, `cholfact!`, and `chol` require that input is either `Hermitian`, `Symmetric`
```
Comment (stevengj) [inline README.md]: added in 0.5

File: README.md (hunk 8 above)
Comment (tkelman) [inline README.md]: the no-op should probably be left in but with a deprecation warning to avoid hard-breaking anyone that was using it

Comment (rofinn) [inline README.md]: For 0.5+ the existing code does:
```julia
macro functorize(f)
    f === :scalarmax       ? :(Base.scalarmax) :
    f === :scalarmin       ? :(Base.scalarmin) :
    f === :centralizedabs2fun ? :(primarytype(typeof(Base.centralizedabs2fun(0)))) :
    f
end
```
Should this be changed to something like:
```julia
macro functorize(f)
    return quote
        Base.depwarn("@functorize is obsolete as of 0.5", Symbol("@functorize"))
        f
    end
end
```
?

**Comment by rofinn** [comment]
Travis failed to download nightly on macos.

**Comment by rofinn** [comment]
Does anyone see any other issues with this?

File: README.md [hunk 13]
```diff
@@ -81,31 +73,12 @@ Currently, the `@compat` macro supports the following syntaxes:
 
 * `@compat Base.IndexStyle(::Type{<:MyArray}) = IndexLinear()` and `@compat Base.IndexStyle(::Type{<:MyArray}) = IndexCartesian()` to define traits for abstract arrays, replacing the former `Base.linearindexing{T<:MyArray}(::Type{T}) = Base.LinearFast()` and `Base.linearindexing{T<:MyArray}(::Type{T}) = Base.LinearSlow()`, respectively.
 
-* `Compat.collect(A)` returns an `Array`, no matter what indices the array `A` has (Julia 0.5 and higher). [#21257]
```
Comment (tkelman) [inline README.md]: this seems recent, it just never worked for 0.4?

File: src/to-be-deprecated.jl [hunk 14]
```diff
@@ -1,18 +1,23 @@
...
 end
 
-if VERSION >= v"0.4.0-dev+3184"
-    include("ngenerate.jl")
-    using .CompatCartesian
-    export @ngenerate, @nsplat
-end
+include("ngenerate.jl")
+using .CompatCartesian
+export @ngenerate, @nsplat
```
Comment (tkelman) [inline src/to-be-deprecated.jl]: should deprecate these

File: src/Compat.jl [hunk 15]
```diff
@@ -690,79 +256,10 @@ end
 
 export @functorize
 macro functorize(f)
```
Comment (tkelman) [inline src/Compat.jl]: should deprecate this

File: README.md (hunk 13 above)
Comment (rofinn) [inline README.md]: I guess this should just remove the (Julia 0.5 and higher part).

File: src/Compat.jl (hunk 15 above)
Comment (rofinn) [inline src/Compat.jl]: Should this be moved to `to-be-deprecated.jl` then?

Comment (tkelman) [inline src/Compat.jl]: yeah good idea, since I don't think we want to keep this around forever

Comment (rofinn) [inline src/Compat.jl]: On a similar note, should we merge `to-be-deprecated.jl` and `ngenerate.jl` into a `deprecated.jl` file?

**Comment by ararslan** [comment]
This looks good to me; I think we should go ahead and merge this. Any objections or further comments?

**Comment by tkelman** [comment]
I'm a little concerned that the deprecations being introduced may never be fixed on 0.5 if packages have moved on to being 0.6-only already. Maybe we should make the deprecation warning 0.6-only since they won't be removed until an 0.6-only version of Compat gets released.

**Comment by rofinn** [comment]
@tkelman Would you be comfortable with an explicit if block in each deprecation?
Example) 
```julia
if VERSION >= v"0.6.0"
    Base.depwarn(...)
end
# code
```

**Comment by tkelman** [comment]
yeah that would be fine imo

File: src/Compat.jl [hunk 16]
```diff
@@ -815,161 +304,14 @@ if !isdefined(Base, :normalize)
...
-if VERSION < v"0.5.0-dev+2285"
-    fieldoffset(T, i) = @compat UInt(fieldoffsets(T)[i])
-    export fieldoffset
-end
-
-if !isdefined(Base, :view)
-    const view = slice
-end
+const UTF8String = Core.String
+const ASCIIString = Core.String
```
Comment (tkelman) [inline src/Compat.jl]: we should probably move these to deprecated.jl, have them const for 0.5, deprecate_binding for 0.6

File: src/Compat.jl [hunk 17]
```diff
@@ -191,169 +119,13 @@ end
...
-        callee.head === :(::) && return true
-        # The other case is with type parameter.
-        # Filter out everything without one.
-        (callee.head === :curly && length(callee.args) >= 1) || return false
-        # Check what the type parameter applies to is a Expr(:(::))
-        return isexpr(callee.args[1], :(::))
-    end
-else
-    new_style_call_overload(ex::Expr) = false
+function new_style_call_overload(ex::Expr)
```
Comment (tkelman) [inline src/Compat.jl]: just in case anyone was calling it from other packages?

**Comment by tkelman** [comment]
Thanks so much for doing this! I think we're good to go here.

File: src/Compat.jl (hunk 17 above)
Comment (rofinn) [inline src/Compat.jl]: Yeah, I'm not sure why they would, but I figured I should try and err on the side of caution.

**Comment by ararslan** [comment]
I'll wait on #380 and rebase it once this is merged. Easier to rebase that than this.

**Pull request merged**